### ISSUE TITLE
Move Unlawful Characters to tests and Do not distribute

### ISSUE THREAD

**Issue by KPrasch** [nucypher/nucypher#2000]
I know, I know enough with the renaming already....

I want to move `unlawful.py` to `tests.characters.unlawful` or similar. Here's why:

- After #1983 - There is an import in `unlawful.py` to `tests.utils.middleware.EvilMiddleWare` which is not packaged in the new `Manifest` and therefore cannot be imported.  This causes all standard installs to crash at import time.   We can wrap the imports and raise an exception, but I find this to be inelegant.
- Unlawful characters are only executed in tests.
- Unlawful characters  are code baggage for *typical* users: there is no compelling reason to distribute unreachable code via `pip install`
- If someone wants to use unlawful entities for security research, we can recommend a developer installation.  

##### Related

#1983, #1005, #367, #1695, #1931

**Comment by tuxxy** [comment]
:+1: I am very much in favor. It's neat having these characters as a demonstration and for testing purposes, but this is merely a _spectacle_ and isn't worth polluting the character API over. It serves no functional purpose to be there and can only contribute crucial, early newcomers.

**Comment by cygnusv** [comment]
I'm also in favor...I've made a similar comment in the past (https://github.com/nucypher/nucypher/pull/1695#discussion_r380152830).

The natural next step after this is to get rid of the `lawful` module, or at least, change it name to something else. This was also discussed in the past (#1005, #367). To this regard, I must admit that I'm very hesitant of doing big renamings now, but it's also a now-or-never situation. Should we open that can of worms?

**Comment by KPrasch** [comment]
> The natural next step after this is to get rid of the lawful module

Not sure if Id go that far - just because they are not siblings does not mean they cannot be conceptually related.  Plus, this implies less changes.

**Comment by jMyles** [comment]
Is it crazy to suggest that we just... package `tests.utils.middleware.EvilMiddleWare`?

**Comment by jMyles** [comment]
(you all suck.)

**Comment by KPrasch** [comment]
Back to my point in the description - I'm not sure I see any value in distributing unreachable code via pip install.

**Comment by jMyles** [comment]
It's not unreachable.  You can reach it with:
`from nucypher.characters.unlawful import Vladimir, Amonia`

Eh?  eh?

**Comment by jMyles** [comment]
OK, here are the arguments I think are worth something:

* First and foremost: the unlawful characters are fun, and fun counts.  People like seeing them in demos, and they make people want to read the codebase.  I realize this is a completely meaningless argument in terms of the cryptographic strength of the library, but socially, there's something there.
* It provides assurance that the actual codebase being used by our users is robust against the attacks described.  Or, maybe there's another plan for that?  I guess we just checksum `nucypher` and call it a day?  I guess that works.
* It gives `lawful` meaning.  The point of `lawful` is, at least in my thoughts, that we can programmatically demonstrate that following the rules is the best way to use the network for a rational actor.  Without `unlawful`, there's no reason to even consider what `lawful` means or how alignment is demonstrated.  As @cygnusv says, it basically warrants deletion if we don't have `unlawful`.
* These are not test Characters.  I mean, it's true that we only use them in tests.  But we have also always ensured that they can survive in a production runtime.  Putting them in the `tests` namespace will mean that they will be the only classes in that namespace that can also work in production (I think?).

At the end of the day, I think that this is just an instinctual and aesthetic preference of mine.  Am I really the only one?

**Comment by cygnusv** [comment]
> * First and foremost: the unlawful characters are fun, and fun counts.  People like seeing them in demos, and they make people want to read the codebase.  I realize this is a completely meaningless argument in terms of the cryptographic strength of the library, but socially, there's something there.

I'm not sure about that. Yeah, it may be fun to _talk_ about them, but I'm not sure if that's the case for an outsider trying to understand our code. See https://github.com/nucypher/nucypher/issues/1005#issuecomment-555607962, for example. And even less comprehensible than `unlawful` are the `chaotic` characters. 

> * It provides assurance that the actual codebase being used by our users is robust against the attacks described.  Or, maybe there's another plan for that?  I guess we just checksum `nucypher` and call it a day?  I guess that works.
If the Imperial College can get away with that, I don't see why we can't too.
Jokes aside, what are you implying? That if tests are not bundled with the released software then there's no assurance that the code is secure? 

> * It gives `lawful` meaning.  The point of `lawful` is, at least in my thoughts, that we can programmatically demonstrate that following the rules is the best way to use the network for a rational actor.  Without `unlawful`, there's no reason to even consider what `lawful` means or how alignment is demonstrated.  As @cygnusv says, it basically warrants deletion if we don't have `unlawful`.

I personally don't think `lawful` has any useful meaning, to be honest. I've talked about this in #1005.

> * These are not test Characters.  I mean, it's true that we only use them in tests.  But we have also always ensured that they can survive in a production runtime.  Putting them in the `tests` namespace will mean that they will be the only classes in that namespace that can also work in production (I think?).

That's a weird criteria. The `_mock_ursula_reencrypts` function in `tests/fixtures.py` allows to produce incorrect re-encryptions from a given Ursula. It's not a fixture, nor it depends on any fixture. Does that make it production material? I really don't think so. It's clearly a test utility.

**Comment by vepkenez** [comment]
> At the end of the day, I think that this is just an instinctual and aesthetic preference of mine. Am I really the only one?

No.

I have always been in favor of doing the right thing about this which is definitely removing unlawful characters from production code...  But I have always thought that maybe for some strange reason, keeping this "cuteness" would increase the odds of success for the network.

Is that crazy?

**Comment by KPrasch** [comment]
To understand unlawful characters as mean-for-production also means recognizing the production use case - which seems to be *security research on the nucypher network*.

Is that really something we want to promote via `pip install nucypher`?
I think no.

If this is the setting, surely wed want that researcher to have a full copy of the codebase to hack on.

**Comment by jMyles** [comment]
Yeah, I agree, @KPrasch.  

On the other hand, it depends how full-throated we want the python API to be.  If people are using it, and importing lawful Characters, then it seems a natural feat of exploration (not so far as security research - just initial poking) to try to import an unlawful Character.  I'd certainly give it a whirl - wouldn't you?  And if people do, then I think it reflects very richly on us.  Few projects have ever put such a thoughtful cryptology forward as to demonstrate robustness against fully codified attacks, available in an identical API to those befitting a rational actor.

But again, this is basically academic.  You are right in everything you're saying.  If the fun aspect isn't enough, then I don't have an argument worth making.

**Comment by KPrasch** [comment]
Perhaps I stand alone in thinking that its still worth preserving the conceptual notion of lawful vs. unlawful vs. chaotic, which I find descriptive and uber-fun.

Do these modules have to be file system siblings for it to be a cohesive API? 
I think no.

**Comment by jMyles** [comment]
I agree they don't *have* to be, but I think that it's better, and prettier, and more likely to draw eyes.  

In fact, I'd rather have `unlawful` adjacent to `lawful` that raises an error on import (telling users to grab the full repo instead of the distribution) than to have it somewhere else where people will only look if they're going out of their way.

**Comment by KPrasch** [comment]
If you are importing from `unlawful` in a sense, you are already going out of your way.

**Comment by vepkenez** [comment]
I mean... we import from `hazmat` here and there don't we?

**Comment by jMyles** [comment]
@KPrasch: My point is this:

Is you are using the python API (and I admit, that may be a small minority, but we certainly still hope for traction), and you, say, auto-import `Alice`, and see `nucypher.characters.lawful` in your import tool, aren't you going to click over there and check out `nucypher.characters.unlawful`?  Of course you are.

Whereas, if you have to grep for `unlawful`, you're probably not going to.  If you see nothing adjacent to `lawful`, it will be more confusing than `lawful` not existing at all (to @cygnusv 's point).

**Comment by KPrasch** [comment]
See related exception from PR #1931 : <https://app.circleci.com/pipelines/github/nucypher/nucypher/4702/workflows/cee52879-a7c4-4427-ae0a-b44ee33eb7b0/jobs/86345/steps>

```
Warning, treated as error:
autodoc: failed to import module 'unlawful' from module 'nucypher.characters'; the following exception was raised:
Traceback (most recent call last):
  File "/home/circleci/.local/lib/python3.7/site-packages/sphinx/ext/autodoc/importer.py", line 32, in import_module
    return importlib.import_module(modname)
  File "/usr/local/lib/python3.7/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 677, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 728, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/home/circleci/.local/lib/python3.7/site-packages/nucypher/characters/unlawful.py", line 33, in <module>
    class Vladimir(Ursula):
  File "/home/circleci/.local/lib/python3.7/site-packages/nucypher/characters/unlawful.py", line 38, in Vladimir
    network_middleware = EvilMiddleWare()
NameError: name 'EvilMiddleWare' is not defined

```

**Comment by jMyles** [comment]
Is that tantamount to saying we won't build docs for anything test-related?

(hanging by a thread here, I know :-) )

**Comment by KPrasch** [comment]
> Is that tantamount to saying we won't build docs for anything test-related?

At the moment yes this True, we are not building docs from docstrings in `tests` (also related to #1931 `Makefile`).

**Comment by jMyles** [comment]
This might be insider blindness, but my sense is that some of the most credibly reassuring docstrings are in `tests`.

**Comment by tuxxy** [comment]
I have to say, if there's any argument that has the ability to change my mind it's going to be the aesthetic argument. Not that I am considering changing my mind... just.. the aesthetics are nice.

**Comment by jMyles** [comment]
Bagh, now I feel bad (observing that #2013 has wrapped the middleware import in a try block).  This is not a hill to die on, and I'm not being insistent.  If the will of the council is to ship `Vladimir` and `Amonia` and whomever else to `tests`, then I'll see it done.  I just want to be sure that I'm really, actually wrong.  :-)
